File: server/common/communication_protocol.py
import socket

def send_message(sock, msg):
    """
    mensaje con formato `<longitud>:<msg>` y espera una respuesta.
    """
    msg_bytes = msg.encode()  
    total_length = len(msg_bytes)
    
    # asegura que se envía todo el mensaje
    sent = sock.send(msg_bytes)
    while sent < total_length:
        sent += sock.send(msg_bytes[sent:])

    return 

def read_message(sock):
    """
    Lee un mensaje con formato `<longitud>:<msg>` desde un socket.
    """
    # Leer hasta encontrar `:` (indica la longitud)
    length_str = b""
    while True:
        byte = sock.recv(1)  
        if byte == b":":
            break
        length_str += byte

    try:
        total_length = int(length_str.decode())  # convertir longitud a entero
    except ValueError:
        raise ValueError(f"Error al convertir la longitud: {length_str}")

    # leo exactamente `total_length` bytes
    received_bytes = b""
    while len(received_bytes) < total_length:
        chunk = sock.recv(total_length - len(received_bytes))
        if not chunk:
            raise ConnectionError("Conexión cerrada antes de recibir el mensaje completo")
        received_bytes += chunk

    return received_bytes.decode()  

File: client/common/communication_protocol.go
package common

import (
	"bufio"
	"fmt"
	"net"
	"strconv"
	"strings"
)

// mensaje con formato `<longitud>:<msg>` y recibe respuesta
func SendMessage(conn net.Conn, msg string) error {
	msgSend := fmt.Sprintf("%d:%s", len(msg), msg)
	//enviar mensaje
	n_write, err := fmt.Fprint(conn, msgSend)
	if err != nil {
		return err
	}

	//asegurarse de que se envió todo el mensaje
	for n_write < len(msgSend) {
		n_write2, err := fmt.Fprint(conn, msgSend[n_write:])
		if err != nil {
			return err
		}
		n_write += n_write2
	}

	//leer respuesta
	return nil
}

// mensaje con formato `<longitud>:<msg>`
func ReadMessage(conn net.Conn) (string, error) {
	reader := bufio.NewReader(conn)
	lengthStr, err := reader.ReadString(':')
	if err != nil {
		return "", err
	}

	//elimino el `:` del final
	lengthStr = strings.TrimSuffix(lengthStr, ":")
	totalLength, err := strconv.Atoi(lengthStr)
	if err != nil {
		return "", fmt.Errorf("error al convertir la longitud: %v", err)
	}

	//Leer exactamente `totalLength` bytes
	bs := make([]byte, totalLength)
	bytesRead := 0
	for bytesRead < totalLength {
		n, err := reader.Read(bs[bytesRead:])
		if err != nil {
			return "", fmt.Errorf("error al leer mensaje completo: %v", err)
		}
		bytesRead += n
	}

	receivedMsg := string(bs[:bytesRead])

	return receivedMsg, nil
}
